### DISCUSSION TITLE
WPA on fallback hotspot

### DISCUSSION THREAD

**Discussion by bobmarley2021** [arendst/Tasmota#18206] [Ideas]
### PROBLEM DESCRIPTION
When losing Wifi connection, Wificonfig option 2 (arguably the most desirable) and Wificonfig option 7 should have the option to use WPA protection.

I am new to Tasmota, but I really like it. Firstly, thanks to all who make this possible. 

I recently purchased a Sonoff S26v2 and flashed it with the latest release (12.4.0). It is connected to a lamp in my daughters bedroom shaped like a unicorn, hence the name "Sonoff-daughtersname-Unicorn-Light". Setting this as the friendly and hostname makes total sense from an administrative point of view as it can be easily found in the router, and identified via web UI. I noticed whilst rebooting my router today, the device falling into hotspot mode and that it was unsecured. Whoops! I thought to myself, I'll login and set that password when my router is back up. One thing that bothered me, is that the SSID is the Hostname (unlike when freshly flashed), so my daughters name was being broadcasted... Hrm... I connected to it when my router was back online and abbreviated the name, see image below. The SSID, however, is the Hostname so now looks ugly in my router. I then looked for the hotspot password option but was surprised that only the UI password exists. I set the UI pass, but when the landing page opened I was still surprised to see my daughters name, because to hide that from everyone in a car outside too, I now have to abbreviate the ~~friendly~~ device name as well. :-S

So I checked the docs:

```
0 = disable Wi-Fi Manager and reboot (used with alternate AP)
2 = set [Wi-Fi Manager](https://github.com/tzapu/WiFiManager/blob/master/README.md#how-it-works) as the current configuration tool and start Wi-Fi Manager (web server at 192.168.4.1) for 3 minutes, then reboot and try to connect Wi-Fi network
4 = retry other AP without rebooting (default)
5 = wait until selected AP is available again without rebooting
6 = Wi-Fi parameters can only be entered via commands in the serial console
7 = set [Wi-Fi Manager](https://github.com/tzapu/WiFiManager/blob/master/README.md#how-it-works) (web server at 192.168.4.1) as the current configuration tool restricted to reset settings only.
```

0 = this is a candidate but there may be times where the router has to be changed (dies?) and the device has no reset button (Shelly1 etc) 
2 = Most convenient but leaks data with using abbreviated names
4 = What if you don't have one?
5 =  same as 0
6 = N/A
7 = this is worse. I don't my neighbours resetting my stuff for me!

I'm not hating on the project, please don't misunderstand me, but isn't this a pretty big flaw? Especially since it's the default option. I looked through the other Wificonfig options, but option 2 makes the most sense or is certainly the most convenient. And isn't the whole point of Tasmota to make it really nice and convenient and easy to use our IoT devices? I was surprised to search for 'wpa' and 'hotspot' in the closed issues and apparently this has not been mentioned before? The closed issues seem to be centred around WPA when Tasmota is connecting to your router. I know there's the argument that the hotspot is only up for 3 minutes. But that is more than enough time for someone to reset it or attack it.

Wouldn't it make sense to just have the option for WPA protection on the hotspot? I just find it odd that this page exists: https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/#securing-your-communication but WPA on the hotspot does not, and more importantly that no password is the default. The option to set the name of the SSID separately would also be good, as then we can leave the hostname alone and just have a non-descript SSID.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Sonoff S26v2
- [x] Tasmota binary firmware version number used: 12.4.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Tasmotizer
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

N/A

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
Wificonfig option 2 or 7 and turn off router ;)

### EXPECTED BEHAVIOUR
For the hotspot to be WPA protected.

### SCREENSHOTS

So yes, I can change the host name and that reflects what the SSID is: 

![wifi](https://tinypic.host/images/2023/03/16/wifi.jpg)

But unless I also change the friendly name, it is shown on the landing page:

![landingpage](https://tinypic.host/images/2023/03/16/landingpage.jpg)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
The default "4" is trying AP1 and AP2 until one of them can be reached. So if you have no AP2 it will try to reach AP1. Best option, no security problem and you can configure an Recovery AP as AP2.

**Reply by bobmarley2021** [reply]
Hi, thanks for replying. Yes I understand what option 4 does but I don't have a second AP for it to connect to. And for AP2 to be the solution, I would have to run a second AP all the time so that if the first one goes down, the second one picks it up before it starts broadcasting the hotspot. Even if I did have a router that could broadcast 2 x 2.4Ghz APs, if the power to the router is down, they're both down. I'm not entirely sure that this is relevant to the issue I have raised? I'm talking about adding WPA encryption to the hotspot... I've just had a quick look and this is possible in WifiManager: https://github.com/tzapu/WiFiManager/blob/master/README.md#password-protect-the-configuration-access-point

Why not just have an option to enable it?

Plus a field to set SSID so the hostname can be different!

**Reply by barbudor** [reply]
You can self compile defining WIFI_AP_PASSPHRASE
Shouldn't this be enough to enable WPA?

I don't think that's something we would like to have by default

**Reply by bobmarley2021** [reply]
I wouldn't have the first idea about compiling it, but I guess I could learn. Is there no way to just have a check box and field in the GUI? Off by default, but easy to turn on via gui or console for compile n00bs like me?

**Reply by sfromis** [reply]
Compiling is easy: https://tasmota.github.io/docs/Compile-your-build/
Simple way of having an extra AP is to enable hotspot on a smartphone.

Anyway, I'd not recommend `WifiConfig 2` due to the ease of access if your AP is pushed offline for whatever reason.

**Reply by Jason2866** [reply]
Still, don't see the problem. The default config does not open an AP. Just tries endless to reach AP1 or AP2. There is NO AP2 as hardware needed.
If you don't want to have to check for AP2 use mode "5". Tasmota tries endless to reach AP1.
Never open a AP (same as Option 4)
Mode 4 and 5 are DIFFERENT from Mode 0!!
I see no need for a change.
Try mode 5 and switch off your wifi network. You won't see any Tasmota AP!

**Reply by bobmarley2021** [reply]
@Jason2866 The default config in this case seems to be 2! 

`{"WifiConfig":{"2":"WifiManager"}}` 

Could that have been triggered by resetting the unit with 6 presses of the button? When I initially linked it up, I typoed my router pass, so I reset with 6 presses. It seems I may have misunderstood 4 and 5. I will keep them in mind if I don't manage to compile it myself. Thank you for the clarification.

@sfromis Thanks for the link, this project looks interesting: https://github.com/benzino77/tasmocompiler docker I _can_ do.

@barbudor could I define WIFI_AP_PASSPHRASE using this step in Tasmocompiler? https://github.com/benzino77/tasmocompiler/blob/master/docs/images/step04.png it says they are included at the end of user_config_override.h

**Reply by Jason2866** [reply]
The default config is WiFi Retry (4)
https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L79

**Reply by barbudor** [reply]
@bobmarley2021

If you have erased the configuration with one of the recovery procedure, and using standard binaries, yes tasmota revert to the AP mode (which is not exactly WifiConfig 2 here because you are in the specific context of a clean configuration without any AP defined. So while default WifiConfig is 4 as Jason wrote, when not home AP is defined, tasmota starts AP mode) 

You are correct and can add a line such as
```
#define WIFI_AP_PASSPHRASE "mysupersecretpassword"
```
In step 4

And you can also take advantage of that to change any of Tasmota défault behavior in case of recovery instead of Restarting from totally blank sheet
- your home SSID and password to avoid reconfiguring it from scratch
- the default WifiConfig if you prefer 2 instead of 4
Etc...

**Reply by bobmarley2021** [reply]
@barbudor Thanks, that's perfect! :)

@Jason2866 yes, I can see that. So how did it get on 2? I had a feeling it was the 6 button press reset. I set wificonfig 4 in console then reset the sonoff with 6 presses. Reconfigured wifi and lo and behold, the wificonfig has been changed to 2!! This should either be well documented or changed, probably the latter. If the rest of the settings are kept with 6 presses except wifi, then surely the wifi config mode should be too? Just like after the initial wifi configuration.

In which case I would like to amend my original issue/discussion to:

1. When resetting wifi settings with 6 presses, after initial wifi reconfig, change wificonfig value back to previous value (or at least default)
and if at all possible:
2. Have a separate SSID field so the host name can be different from the SSID

edit: Sorry @barbudor, I didn't see your comment about resetting procedures and it being in AP mode which is not strictly 2. When you say AP mode, do you mean as in after an initial flash? In which case it sets to wificonfig 4. It doesn't seem to do this after a reset though.
